MazerPlatformer/Character.cs

```diff
@@ -44,17 +47,17 @@ public enum CharacterDirection { Up, Down, Left, Right };
         public virtual event CollisionDirectionChanged OnCollisionDirectionChanged;
 
         /* Delegates that are used for the character events */
-        public delegate void DirectionChanged(CharacterDirection direction);
-        public delegate void CollisionDirectionChanged(CharacterDirection direction);
-        public delegate void StateChanged(CharacterStates state);
+        public delegate Either<IFailure, Unit> DirectionChanged(CharacterDirection direction);
+        public delegate Either<IFailure, Unit> CollisionDirectionChanged(CharacterDirection direction);
+        public delegate Either<IFailure, Unit> StateChanged(CharacterStates state);
 
         protected Character(int x, int y, string id, int width, int height, GameObjectType type, int moveStepIncrement = 3) :
             base(x, y, id, width, height, type)
         {
             _moveStep = moveStepIncrement;
         }
 
-        public override void Initialize()
+        public override Either<IFailure, Unit> Initialize()
         {
             base.Initialize();
```

Comment:
You can't be sure that external libraries/code will not throw exceptions - These can be suppressed by wrapping the functions with Ensure() which will ensure that any unexpected exceptions thrown will be converted to failures and therefor never leave the function and travel up the event handler heighrachy.  This is problematic for two reasons:

- Your function will not be guaranteed to return its result.
- You don't know if the exception handling code takes you out of the normal operating procedure of your logic and how adequate it will be at resolving or resuming the code. At least with ensuring it **always returns**, the caller always has the ability to implement failure code. And in particular if it always returns an `Either<IFailure, Unit>` then you must either make provision for the failure or it will automatically cancel your pipeline.